### ISSUE TITLE
Restore Producing Unreliable Results & Not Handling 'Unusual' Directory Names Well?

### ISSUE THREAD

**Issue by madranet** [gilbertchen/duplicacy#346]
An elderly relative of mine recently spilled liquid on her laptop and killed it. Luckily I'd installed Duplicacy on it and made a backup to S3, just before it died so, when she got a new laptop,  I was able to do a `duplicacy restore -r X` and get all her "stuff" back. 

Or so I thought...

A few days after I'd done the restore, she contacted me to say that there were several files missing. In each case, the missing files were the contents of subdirectories under her main `~/Documents` directory.  However not all. Some subdirectories under `~Documents` restored fine, others just restored the subdirectory itself [and in some cases sub-subdirectories] but with no files inside.

When I checked the backup archives with `duplicacy list -files` I could see that most of these directories were actually present in the S3 backup but, for whatever reason, the directories had restored with no content. Other subdirectories however were completely missing from the backup archives, so had not even backed up in the first place.

The filters for the backup in question are extremely simple, so there doesn't seem to be a question of misconfiguration, unless i'm missing something really obvious:

```
+Documents/*

+Pictures/*

-*
```

I did some further digging and found possible reasons why some of the directories had not backed up or restored:

* One of the directories which had backed up but did restore had a `/` character in the directory name : `songs/stories` This is an acceptable directory name on OSX where the path separator is `:` but obviously when backed up to a linux server [S3] that `/` character was treated as a path separator and so caused Duplicacy to lose track of it. 

I'm not sure what could be done about that situation and I'm not saying it's Duplicacy's fault, but an odd 'gotcha' to be aware of.

* A couple of the other directories had trailing spaces in their names [as I said, this was an elderly relative's computer, so some of the naming practices were a bit odd!] For example, one directory apparently called `Letters` actually turned out to be called `Letters<space> `. Those directories had also backed up to the archive but were missing from the restore. So it seems like somewhere along the line Duplicacy ignored a trailing space it had previously honoured and that caused those directories not to restore.

* Another directory which backed up but failed to restore was entitled `Health,Food,Tabs`.  On the face of it nothing wrong with that as a directory name on any platform. But possibly Duplicacy got confused by the commas?

Just to add to the confusion, there were a few other directories where I couldn't see anything odd about the naming at all but which still failed to restore.

In the end, I was able to restore all the content of each of the missing directories by using individual `duplicacy restore  -r X +"path/to/directory/*"` commands. So it all worked out, albeit a bit more painstakingly than I'd have liked.  

But it has left me a bit concerned on how reliable Duplicacy's `restore` command is.  Admittedly there were some unusual directory naming conventions in that backup and the ones involving the use of `/`  were doubtless bound to cause trouble. But surely Duplicacy should be able to handle something like a trailing space or a comma in a directory name without that causing files to fail to be restored?  And what about the contents of the other directories which had nothing apparently odd about their names, but which [seemingly] randomly failed to restore?

Unfortunately also, as with at least [one other issue](https://github.com/gilbertchen/duplicacy/issues/246) I've run into when using Duplicacy, there was nothing in the output from the original `duplicacy restore` command to indicate that there'd been any problems or that any files couldn't be found and hadn't been restored. Where there was a problem [for whatever reason] with one of these subdirectories, Duplicacy just restored it as an empty directory and then went on its merry way, without reporting that anything had gone amiss.

**Comment by gilbertchen** [comment]
I ran a quick test of backing up and restoring a few directories containing `:`, ` `, and `;`.  All worked fine.  Which version are you running?  At least 2.0.10 doesn't have this issue according to my test.

> One of the directories which had backed up but did not restore had a / character in the directory name : songs/stories This is an acceptable directory name on OSX where the path separator is : 

I can't create a directory that contains a `/` either in Terminal or Finder.  In Finder you can enter `:` in a folder name which will be converted into `/` when displaying, but in Terminal the name is still `:`.

**Comment by jonreeves** [comment]
@gilbertchen out of curiousity, how does duplicacy react to restoring a filename thats incompatible with the destination machine... Was initially from osx to a windows machine? e.g.. if a file had a colon in it, `file:name.txt` does it become `file_name.txt` or something?

**Comment by madranet** [comment]
>I ran a quick test of backing up and restoring a few directories containing :, , and ;. All worked fine. Which version are you running? At least 2.0.10 doesn't have this issue according to my test.

I'm not sure exactly which version. The original backup was made at the end of November, so whichever version was current then. And the initial restore a couple of weeks ago, so probably 2,0,10 for that.

>I can't create a directory that contains a / either in Terminal or Finder. In Finder you can enter : in a folder name which will be converted into / when displaying, but in Terminal the name is still :

Which version of OSX are you on?  On El Capitán I can create directories with `/` in their name in the Finder:

![2018-01-31_10-07-32](https://user-images.githubusercontent.com/669658/35616912-94993340-066e-11e8-9b77-a503fe7fdfb5.png)

But, in the Terminal, they are shown as having colons:

![2018-01-31_10-08-14](https://user-images.githubusercontent.com/669658/35619997-9e1278a6-0677-11e8-8555-16d3c098ac94.png)

Trying to create a directory with `/` in the name via Terminal results [as expected] in a `No such file or driectory` error [unless the `-p` flag is used, which creates a directory heirarchy]

Trying to create a directory with `:` n the name via Terminal results in the `:` being replaced with a `/` when viewed in the Finder:

![2018-01-31_11-13-55](https://user-images.githubusercontent.com/669658/35620179-3fe278c0-0678-11e8-9889-316bcadbc863.png)

![2018-01-31_11-17-32](https://user-images.githubusercontent.com/669658/35620208-5c6313b0-0678-11e8-8f02-5aa3434d4b4e.png)

All of which is interesting and indicative of the schizophrenic nature of OSX but not directly related to Duplicacy --although @jonreeves does raise an interesting point as to how Duplicacy should deal with these situations.

Anyway,  I decided to do a test myself. So I created a hierarchy of files and folders with dodgy directory names:

![2018-01-31_11-24-08](https://user-images.githubusercontent.com/669658/35624016-8dfcaba4-0685-11e8-8fc9-bc4094715187.png)

and initialised a duplicacy archive and backed up to S3.  

```
★ Dalek ★  test  ᐅ  duplicacy init -encrypt test s3://eu-west-1@amazon.com/xxxxxxxxxxxx/test/
Enter S3 Access Key ID:XXXXXXXXXXXX
Enter S3 Secret Access Key:XXXXXXXXXXXXXXXXXXXX
Enter storage password for s3://eu-west-1@amazon.com/xxxxxxxxxxxx/test/:********
Re-enter storage password:********
/Users/madra/Desktop/test will be backed up to s3://eu-west-1@amazon.com/xxxxxxxxx/test/ with id test

★ Dalek ★  test  ᐅ  duplicacy backup
Storage set to s3://eu-west-1@amazon.com/xxxxxxx/test/
Enter storage password:********
No previous backup found
Indexing /Users/madra/Desktop/test
Listing all chunks
Packed .DS_Store (6148)
Packed some:other:stuff/test01.txt (631)
Packed some:other:stuff/test02.txt (2524)
Packed this:that:the other/test03.txt (1132)
Packed this:that:the other/test04.txt (832)
Packed thngs,stuff,bits/test05.txt (596)
Packed thngs,stuff,bits/test06.txt (0)
Packed trailing space /test07.txt (1012)
Packed trailing space /test08.txt (775)
Backup for /Users/madra/Desktop/test at revision 1 completed
```

Then I deleted the folder from my desktop, recreated it and did a `duplicacy restore r 1` and [wouldn't you know it!] everything restored perfectly, with no issues:

```
★ Dalek ★  test  ᐅ  cd ~/Desktop && mkdir restoretest && cd restoretest
★ Dalek ★  restoretest  ᐅ  duplicacy init -encrypt test s3://eu-west-1@amazon.com/xxxxxxt/test/
Enter S3 Access Key ID:XXXXXXXXXXXX
Enter S3 Secret Access Key:XXXXXXXXXXXXXXXXXXXX
Enter storage password for s3://eu-west-1@amazon.com/xxxxxxxx/test/:********
The storage 's3://eu-west-1@amazon.com/xxxxx/test/' has already been initialized
Compression level: 100
Average chunk size: 4194304
Maximum chunk size: 16777216
Minimum chunk size: 1048576
Chunk seed: c0229c8d6be60XXXXXXXXXXXXXXX7d702a7a3d438a6cc72f1ada8deef

★ Dalek ★  restoretest  ᐅ  duplicacy restore -r 1
Storage set to s3://eu-west-1@amazon.com/xxxxxxxx/test/
Enter storage password:********
Restoring /Users/madra/Desktop/restoretest to revision 1
Downloaded .DS_Store (6148)
Downloaded thngs,stuff,bits/test06.txt (0)
Downloaded some:other:stuff/test01.txt (631)
Downloaded some:other:stuff/test02.txt (2524)
Downloaded this:that:the other/test03.txt (1132)
Downloaded this:that:the other/test04.txt (832)
Downloaded thngs,stuff,bits/test05.txt (596)
Downloaded trailing space /test07.txt (1012)
Downloaded trailing space /test08.txt (775)
Restored /Users/madra/Desktop/restoretest to revision 1
Total running time: 00:00:01
```

![2018-01-31_12-51-58](https://user-images.githubusercontent.com/669658/35624069-c961a0f0-0685-11e8-85c5-90b6f6c18828.png)

So now I'm baffled. 

On the one hand, this is great news as it means I was unable to trip Duplicacy up by adopting strange naming conventions. On the other hand it suggests that *something else* caused the contents of certain folders in my original backup to either not upload to Amazon S3 in the first place or fail to download when a `restore` command was issued.

In a way it would have been easier if the problem was down to unfortunate naming conventions, as that would have been simpler to track down or avoid. But, unfortunately it seems it's going to be one of those weird voodoo bugs that only rears its head in some really unusual set of circumstances.

**Comment by tophee** [comment]
> In the end, I was able to restore all the content of each of the missing directories by using individual duplicacy restore -r X +"path/to/directory/*" commands. So it all worked out, albeit a bit more painstakingly than I'd have liked.

Just to clarify: are you saying that, in the end, no data at all was lost, despite the fact that

> Other subdirectories however were completely missing from the backup archives, so had not even backed up in the first place.

So it was just luck that there were no files in those folders, or what?

**Comment by gilbertchen** [comment]
If Duplicacy doesn't back up a directory, it is most likely due to permission issues for which it will give warnings like list:
```
Subdirectory dir1/ cannot be listed
Subdirectory dir2/ cannot be listed
File file1 cannot be opened
Backup for /Users/chgang/AcrosyncTest/repository at revision 2 completed
2 directories and 1 file were not included due to access errors
```

Earlier versions of Duplicacy do not back up empty directories.  This was fixed by https://github.com/gilbertchen/duplicacy/commit/22ddc04698077fc297ae7408991405fc4aac0fad

> how does duplicacy react to restoring a filename thats incompatible with the destination machine

Duplicacy doesn't check if the filename is legitimate on the destination machine.  Instead, it will just attempt to create the file and bail out if there is any error.

**Comment by madranet** [comment]
> Just to clarify: are you saying that, in the end, no data at all was lost, despite the fact that
> 
>> Other subdirectories however were completely missing from the backup archives, so had not even backed up in the first place.
> 
> So it was just luck that there were no files in those folders, or what?

@tophee Whoops! --I should have elaborated on that in my original post.  

What happened was that, subsequent to  the original partial restore, when I found out some of the directories hadn't backed up in the first place, I was posted the hard drive extracted from the dead laptop and was able to mount it and run more `duplicacy backup`s.  

For some of the directories which still refused to back up, I had temporarily to explicitly add that directory to the `filters` file [even though as a subdirectory of `~/Documents` it should have been included automatically] and then run the back up again.  In this way I was able to eventually force those directories to back up from the salvaged hard drive.  

It was after completing this step that I was able to connect to the new laptop remotely and initiate the restores at that end, using `individual duplicacy restore -r X +"path/to/directory/*"` where needed to force the recalictrant directories to populate.

So, yes, the end result was that I got all the files back but I forgot to mention that I was able to make subsequent backups to fill in the gaps missing from the original ones.

>If Duplicacy doesn't back up a directory, it is most likely due to permission issues for which it will give warnings like list:

There definitely wouldn't have been any permissions problems.  All the missing documents were just word processor docs, images, etc saved inside the person's `~/Documents` folder. The user in question wouldn't even know the command line existed, never mind somehow be able to change permissions on a file.

**Comment by TowerBR** [comment]
> So, yes, the end result was that I got all the files back but I forgot to mention that I was able to 
> make subsequent backups to fill in the gaps missing from the original ones.

Ah, now it's clearer! So in the end, the restore worked, "in pieces."

> > If Duplicacy doesn't back up a directory, it is most likely due to permission issues for 
> > which it will give warnings...

> There definitely wouldn't have been any permissions problems. 

But this question continues...

**Comment by gilbertchen** [comment]
> For some of the directories which still refused to back up, I had temporarily to explicitly add that directory to the filters file [even though as a subdirectory of ~/Documents it should have been included automatically] and then run the back up again. In this way I was able to eventually force those directories to back up from the salvaged hard drive.

@madranet what are these directories like?  Can you run `duplicacy -d backup` with a `filters` file that doesn't include any of these directories to see why they are excluded?

**Comment by madranet** [comment]
>what are these directories like?

@gilbertchen just your everyday directories filled with various word processor docs and the odd picture file. Nothing odd or unusual

> Can you run duplicacy -d backup with a filters file that doesn't include any of these directories to see why they are excluded?

I'm trying to test this, using `duplicacy -d  backup > ~/Desktop/backup.txt` but when I inspect the file produced, it's just listing all the directories and sub-directories involved, not their contents:

```
<snip>
Listing Documents/Songs General/
Listing Documents/Stories & Songs/
Listing Documents/Stuff Left in Downloads/
Listing Documents/Stuff Left in Downloads/Coming Up for Air/
Listing Documents/Stuff Left in Downloads/Reunion marking/
Listing Documents/Stuff Left in Downloads/Wool Omnibus Edition/
Listing Documents/Stuff Left in Downloads/Wool Omnibus Edition (2285)/
Listing Documents/WW1 and 2/
Listing Documents/War Deaths.rtfd/
Listing Pictures/
Listing Pictures/Photos Library.photoslibrary/
<snip>
```
so it's not  much good for finding out whether or not any of the directories would have been uploaded empty. Also bear in mind that this external hard drive has now been 'modified' in that I went through and fixed all the odd naming conventions I found [when I thought that might be the problem]. So what I'm seeing here is not necessarily what happened during the original backup/s.

Just for interest's sake, here is the output of using the `history` flag.  The first dozen revisions were the original backups, run on the now defunct laptop. The subsequent revisions are the ones where I was working on the salvaged hard drive, trying to  recover the missing things.

First a directory with nothing apparently wrong with its filename:

```
★ Dalek ★  ᐅ  duplicacy history "Documents/Research"
Storage set to s3://eu-west-1@amazon.com/xxxxxxxxxxx
      1:
      2:
      3:
      4:
      5:
      6:
      7:
      8:
      9:
     10:
     11:
     12:
     13:
     14:
     15:
     16:
current:            5032 2018-01-30 14:54:31        Documents/Research
```
and now one of the Directories with a dubious name:

```
★ Dalek ★  ᐅ  duplicacy history "Documents/Stories/Songs"
Storage set to s3://eu-west-1@amazon.com/xxxxxxxxxx
      1:
      2:
      3:
      4:
      5:
      6:
      7:
      8:
      9:
     10:
     11:
     12:
     13:
     14:
     15:
     16:
current:
```
..and just in case the directory name got converted to Apple command line style separators:

```
★ Dalek ★  ᐅ  duplicacy history "Documents/Stories:Songs"
Storage set to s3://eu-west-1@amazon.com/xxxxxxxxxx
      1:
      2:
      3:
      4:
      5:
      6:
      7:
      8:
      9:
     10:
     11:
     12:
     13:
     14:
     15:
     16:
current:
```
[The reason that particular directory isn't present in any backups including the newest ones, is because, after revision 13 or so,  I changed its name to `Stories & Songs` while I was pursuing the theory of the naming conventions being at the root of the problem.

It's all very strange. I've been unable to reproduce this at all and now I've 'modified' the salvaged hard drive by renaming the dubiously named directories in it, everyting is backing up and restoring fine. yet, for whatever reason, certain directories didn't back up  and others, which had backed up restored devoid of content

One more [seemingly random, but you never know!] factor that just occurred to me as I mounted the drive again to run the above test: 

The drive is locked with open firmware. So I need to put in a password when mounting it as an external drive.  However, as it would already have been mounted when in use on the defunct laptop, I can't see how that could have effected anything in the operation of Duplicacy on that laptop.  But just thought I'd mention it since sometimes these weird edge cases can lead to some unexpected gotchas.

**Comment by madranet** [comment]
Just been twiddling again here and I may be misusing the `history` flag.  

If I run `history` on one of the directories which restored empty, I get an empty list, until the last revision:

```
★ Dalek ★  ᐅ  duplicacy history "Documents/Research"
Storage set to s3://eu-west-1@amazon.com/xxxxxxxxxxx
      1:
      2:
      3:
      4:
      5:
      6:
      7:
      8:
      9:
     10:
     11:
     12:
     13:
     14:
     15:
     16:
current:            5032 2018-01-30 14:54:31        Documents/Research
```

However, if I use `history` on one of the documents within that directory, then I can see that that document was present from the first backup --but, for whatever reason it didn't restore, when its containing directory did [I've removed the chunk codes [?] for brevity]:

```
★ Dalek ★  ᐅ  duplicacy history "Documents/Research/3rd Ypres Langemarc.odt"
Storage set to s3://eu-west-1@amazon.com/xxxxxxxx
      1:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
      2:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
      3:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
      4:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
      5:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
      6:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
      7:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
      8:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
      9:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
     10:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
     11:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
     12:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
     13:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
     14:           35477 2016-06-30 11:09:42  Documents/Research/3rd Ypres Langemarc.odt
     15:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
     16:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
current:           35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
```

With the directories which didn't back up in the first place, obviously, neither a `history` on the directory or on a file it should have contained produce anything.

**Comment by TowerBR** [comment]
> I'm trying to test this, using duplicacy -d backup > ~/Desktop/backup.txt

@madranet maybe it's worth a try:

```duplicacy list -storage %storage_url% -files > ~/Desktop/backup.txt```

or just:

```duplicacy list -files > ~/Desktop/backup.txt```

**Comment by gilbertchen** [comment]
@madranet  thank you for providing all the details.  This is perhaps my last question: could those files under `Documents/Stories:Songs` appear in the backups under a different directory name?  

This command can find it out:
```
duplicacy list -files -r 1-12 | grep <filename>
```

**Comment by jonreeves** [comment]
I have a thought but what could have happened, it's a bit out there, but someone with more MacOS experience may beable to confirm or deny my suspicions...

I'm wondering if those folders that didn't backup did that old `greyed out` thing where the folder gets set to `Jan 1984`. I think this happens when the system detects there is an issue with the files/folder and typically you cannot open or access the item(s) in question. To fix it you typically have to copy/move or change the date on the folder, I think renaming it works too.

If the user hadn't interacted with those folders for a while, it would be possible to have not noticed that they couldn't access them. Additionally when you connected it as an external HDD and renamed the them you might have fixed what was wrong with them (allowing them to suddenly backup).

The reason I believe its plausible is that I recently (last friday) saw Backblaze Unlimited get stuck on a friend's mac, saying it still had X number of folders to backup but never progressing (for days). When we saw what folder it was stuck on, it was `greyed out`, so she copied it to another folder and deleted the original. The backup then continued with no issues.

Again, I'm not a MacOS user, I've just seen and read about this `grey out` thing before and thought it could be the culprit in this scenario. I don't know if or how you can induce it yourself, but perhaps someone with some better knowledge of Macs can take it from here?

**Comment by madranet** [comment]
@jonreeves  --At this stage of the proceedings, I'm open to any suggestions, no matter how 'off the wall' they sound! 

I've got to say though that I've never noticed this 'greying out' happening on any version of OSX I've used and the person in question didn't have a huge archive of material. She tends to just save everything in `~/Documents` [that is the ones she doesn't leave on the Desktop or in her Downloads folder --but that's another story for another day!]. I could see where maybe one or two of the sub-folders might not have been opened for a while, but several of them contained a lot of 'stuff' she'd been working on recently. And I'm sure I'd have got an SOS text message if they'd been behaving oddly at all.

@gilbertchen & @TowerBR 

Well, the plot thickens.  I've done some dumps of the `-files` at various revisions and `diff`-ed them in [TextMate](https://github.com/textmate/textmate) and I'm not seeing those directories listed at all in any of the revisions. 

Listing from `r 1`. Note `Documents/Research/` subdirectory does not exist

```
<snip>
122782 2017-08-15 15:05:18 Documents/Remote instruct.odt
   33995 2015-07-15 10:00:33 Documents/Remove network.odt
   33199 2017-09-23 19:20:50 Documents/Screen Shot 2017-09-23 at 19.20.39.png
   13502 2016-12-14 19:52:57 Documents/Seven Wonders of the World.odt
   84817 2014-05-20 14:34:42 Documents/Sky Debacle.odt
   10127 2017-02-20 13:08:44 Documents/Sky address.odt
<snip>
```

Listing from `r 7`. Still no record of `Documents/Research/` subdirectory

```
<snip>
   122782 2017-08-15 15:05:18 Documents/Remote instruct.odt
   33995 2015-07-15 10:00:33 Documents/Remove network.odt
   33199 2017-09-23 19:20:50 Documents/Screen Shot 2017-09-23 at 19.20.39.png
   13502 2016-12-14 19:52:57 Documents/Seven Wonders of the World.odt
   84817 2014-05-20 14:34:42 Documents/Sky Debacle.odt
   10127 2017-02-20 13:08:44 Documents/Sky address.odt
<snip>
```

Yet, if I use @gilbertchen' s suggestion and look for one particular file within that subdirectory across revisions, it's there every time, where it's supposed to be: in the same subdirectory that is missing from the listings produced by `-files`. It's like it exists and also doesn't exist at the same time ???

```
★ Dalek ★  ᐅ  duplicacy list -files -r 1-7 | grep "3rd Ypres Langemarc.odt"
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt
```

*[When comparing timestamps on the `files` listings, I realised that the last backup made from the original laptop was `-r 7` not `r 12` as I'd thought.  Hence I only grepped across 1-7.]*

I, for one, am officially baffled!

**Comment by gilbertchen** [comment]
` duplicacy list -files` doesn't list directories.  If you run `duplicacy list -files -r 1-7| grep <filename>` for a file that you know exist under `Documents/Stories:Songs` would that come up with anything?

**Comment by madranet** [comment]
>duplicacy list -files doesn't list directories

Ah. I didn't know that. But shouldn't it still have listed the files within that directory –even if it didn't list the directory itself?

>If you run duplicacy list -files -r 1-7| grep <filename> for a file that you know exist under Documents/Stories:Songs would that come up with anything?

No apparent problem there either:

```
★ Dalek ★  ᐅ  duplicacy list -files -r 1-7 | grep "A Vision of Judgment.odt"
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
```

**Comment by TowerBR** [comment]
> It's like it exists and also doesn't exist at the same time ???

I also didn't understand...

> But shouldn't it still have listed the files within that directory –even if 
> it didn't list the directory itself?

I also think it would be the most correct.

I think the list should look like this:

```
<snip>
   122782 2017-08-15 15:05:18 Documents/Remote instruct.odt
   33995 2015-07-15 10:00:33 Documents/Remove network.odt
   35477 2016-06-30 11:09:42 Documents/Research/3rd Ypres Langemarc.odt   <<<<====
   33199 2017-09-23 19:20:50 Documents/Screen Shot 2017-09-23 at 19.20.39.png
   13502 2016-12-14 19:52:57 Documents/Seven Wonders of the World.odt
   84817 2014-05-20 14:34:42 Documents/Sky Debacle.odt
   10127 2017-02-20 13:08:44 Documents/Sky address.odt
<snip>
````

**Comment by TowerBR** [comment]
Since we are evaluating "out of the wall" possibilities, let me make another guess: the other day I was testing Duplicacy with Google Drive, uploading to a remote folder (let's say "folder1"), but when I checked there via web interface those files were not there! But I could list them, restore, etc. 

Then I did a search using the web interface using file names and found them, they were marked as "orphaned". To summarize, I had deleted the original folder and changed it to a similar one (say "folder_1"), where the files obviously were not, but Google Drive still "viewed" the original "folder1" folder and was accepting the Duplicacy upload, but tagged the files as "orphaned".

In short: the files were there, but they were not in any folder.

Could something similar have happened?

**Comment by gilbertchen** [comment]
> ★ Dalek ★  ᐅ  duplicacy list -files -r 1-7 | grep "A Vision of Judgment.odt"
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt
   30132 2015-08-26 10:36:24 Documents/Stories:Songs/A Vision of Judgment.odt

So the files under `Documents/Stories:Songs` did get backed up.  I was under the impression that they didn't.

Do you know any other directories whose files didn't get backed up?  I.e., they can't be found by `duplicacy list -files -r 1-7| grep <filename>`.

**Comment by madranet** [comment]
Well, that's the thing. I *thought* the files hadn't been backed up because I'd been unable to list them.  See previous comments re nothing being listed under ~/Documents/Research`

But now that I've been searching for individual files by name, it seems they *are* backed up. Yet they only appear in file listings if I explicitly ask for them by name.  And when a `restore` was initiated, only the folders that contained them were restored, not any of the files themselves.

Like I said before, it's like they exist and don't exist at the same time.

Re: @TowerBR's Google Drive comment: these files were just 'regular' Finder files. Backup was to Amazon S3.

**Comment by madranet** [comment]
You're not going to believe this, folks!

I've just dug one of the original `-files` listings out of the bin and had a search  through it for one of the missing directories `Documents/Research` and, as before:

```
<snip>
122782 2017-08-15 15:05:18 Documents/Remote instruct.odt
33995 2015-07-15 10:00:33 Documents/Remove network.odt
33199 2017-09-23 19:20:50 Documents/Screen Shot 2017-09-23 at 19.20.39.png
13502 2016-12-14 19:52:57 Documents/Seven Wonders of the World.odt
84817 2014-05-20 14:34:42 Documents/Sky Debacle.odt
<snip>
```

Since it's only on line 70 of the file, i've just been looking directly at that location in the file, rather than searching the whole 70000 lines. So, on a whim, I decided to search for `Documents/Research` and, lo' and behold, way down at line 500+ I find:

```
<snip>
   11302 2016-12-02 15:52:46 Documents/Research/12th Batt database.odt
   16567 2016-07-01 15:26:04 Documents/Research/12th Batt list.odt
   16002 2016-07-01 15:54:52 Documents/Research/12th Batt local soliders killed.odt
   61170 2015-01-03 11:24:54 Documents/Research/1901C Jam.odt
  108457 2013-05-20 16:11:08 Documents/Research/1907 Dir Fam.odt
<snip>
```
with the other 'missing' directories and their contents down beyond line 700 somewhere.

So, it looks Duplicacy processes each subdirectory after it has run through all the files in the top level directory and thus, the output from `-files` is not in alphabetical order --as it looks on first glance-- but in hierarchical order.

*\<insert facepalm smiley!\>*

But, before we get too carried away; while this solves "The mystery of the Files Which Were There But Not There" [one of Conan-Doyle's lesser known Sherlock Holmes tales], it still doesn't answer the mystery of why certain subdirectories restored bereft of content.

**Comment by TowerBR** [comment]
Things are getting better!   ;-)

**Comment by jeremejevs** [comment]
@madranet Have you managed to do a complete restore? Or did you just copy the contents off the old hard drive and move on? I'm currently evaluating Duplicacy, and would like to be sure that I won't have trouble restoring in the future (doing drill restores is still an option, but can get expensive with storage providers which charge you for egress, like B2).

**Comment by jeremejevs** [comment]
@madranet Thanks for reporting back! So, it's possible there would be no happy ending if that drive were to die with that laptop.

Decided to try Duplicacy anyway though, there's a story like this with any tool. Just going to also have a second local backup, which I can test more frequently than a remote one.

**Comment by madranet** [comment]
No worries. Sorry I couldn't be of more help.

And I'm still using Duplicacy myself. Though thankfully I've not had to avail of the restore function again, thus far. Touch wood!
